web3/_utils/request.py

```diff
@@ -71,21 +77,18 @@ async def cache_async_session(endpoint_uri: URI, session: ClientSession) -> None
             for key, session in evicted_items.items():
                 await session.close()
 
+_session_cache = SessionCache(size=8)
+_async_session_cache = SessionCache(size=8)
 
-def _remove_session(key: str, session: requests.Session) -> None:
-    session.close()
-
-
-_session_cache = lru.LRU(8, callback=_remove_session)
 _async_session_cache_lock = threading.Lock()
-_async_session_cache = SessionCache(size=8)
+_session_cache_lock = threading.Lock()
 
 
 def _get_session(endpoint_uri: URI) -> requests.Session:
     cache_key = generate_cache_key(endpoint_uri)
     if cache_key not in _session_cache:
-        _session_cache[cache_key] = requests.Session()
-    return _session_cache[cache_key]
+        cache_session(endpoint_uri, requests.Session())
+    return _session_cache.get_cache_entry(cache_key)
```

Comment:
This cache access should be protected by the lock as well, IMO.

Reply:
I don't disagree with you. I think I tried that in the begining. I feel like I started getting deadlocks for some reason. My thought process was as long as everyone stops while updating the cache then it should be ok to read the cache in parallel. 

Any chance you could try this code out and see if it works in your process? It would be nice to get a real world test of it.

Reply:
I didn't notice that `lru.LRU` is gone. Since the GIL is not dropped by a C extension anymore, the code might be thread-safe now even without locks. But still, I recommend adding a lock as calls to `requests.Session` methods may drop the GIL now or in the future.

Such a simple locking works for us at the moment: https://github.com/ethereum/web3.py/issues/1847#issuecomment-1086109749

Reply:
@Eenae I added the lock you suggested in the latest commit. My test seemed to work this time. Do you mind testing it out?

Reply:
@dbfreem thanks. I did some research and have the following points to highlight:

1. A code fragment which requires exclusive access to a data structure must not release a lock until the work is done. I mean acquiring the lock in multiple touches is not correct since there may be concurrent access between them. And any access should be protected by a lock, including the `in` lookup. IMO these lines must be protected by a single `acquire` https://github.com/ethereum/web3.py/blob/6c845b6eddcff2f2180829b0084fc63e0aae604b/web3/_utils/request.py#L89-L92

2. That leads us to the second point. Now I suspect what might be the reason of deadlocks you mentioned. We can't just wrap the lines above with the lock since it'll stumble upon the second `acquire` in `cache_session`. That's because `Lock` is not reentrant. You might re-arrange the code or use `threading.RLock` to overcome this.

3. Also I remembered about a note here https://docs.python.org/3/library/threading.html#threading.Lock.acquire : "Changed in version 3.2: Lock acquisition can now be interrupted by signals on POSIX if the underlying threading implementation supports it.". IMO that means it's not safe to ignore a return value of `acquire`, meaning that it might not be safe to use `Lock` or `RLock` as a context manager as-is. I'd rewrite it with a helper:
```python
@contextmanager
def locked(lock):
    """
    Contextmanager which locks and releases a lock in the safest way
    :param lock: Lock or RLock
    """
    while not lock.acquire():   # may be interrupted by a signal, so a loop is needed
        pass

    try:
        yield
    finally:
        lock.release()
```

At the moment we're not on the `master` even not on the current version in production, so I'll not be able to try out this PR, unfortunately.

Reply:
Ok, so I rearranged the code to put the locking all in one method. So now all access to the session cache should have a lock around it. I did stick with the out of the box python context manager of the lock since python says it can be used that way. 
[python thread docs context manager](https://docs.python.org/3/library/threading.html#using-locks-conditions-and-semaphores-in-the-with-statement)

Thanks for all the reviews @Eenae you made me think about how this actually needs to flow and consolidate it down to a single method for blocking and async session caches